File: functionality/functionality_executor.py
import logging
import base64
import tempfile
import os
import threading
from typing import Dict, List, Callable, Any, Optional
import concurrent.futures
import time

from agent.component import Component
from event.event import Event
from event.event_type import EventType
from event.event_dispatcher import EventDispatcher
from communicator.lapras_topic import LaprasTopic
from communicator.message_type import MessageType
from communicator.mqtt_communicator import MqttCommunicator
from functionality.functionality_invocation import FunctionalityInvocation
from functionality.functionality_signature import FunctionalitySignature
from util.data_type import DataType


class FunctionalityExecutor(Component):
    """
    Executes functionality methods that are registered with the system.
    Handles incoming functionality requests via MQTT and manages their execution.
    """
    
    def __init__(self, event_dispatcher: EventDispatcher, agent):
        super().__init__(event_dispatcher, agent)
        self.logger = logging.getLogger(self.__class__.__name__)
        
        # Executor for running functionalities
        self.functionality_executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
        
        # Maps for storing functionality handlers and signatures
        self.functionality_function_map: Dict[str, Callable] = {}
        self.functionality_signature_map: Dict[str, FunctionalitySignature] = {}
        
        # Get MQTT communicator and agent name from agent
        self.mqtt_communicator = agent.get_mqtt_communicator()
        self.agent_name = agent.get_agent_config().get_agent_name()
    
    def subscribe_events(self):
        """Subscribe to relevant events"""
        self.subscribe_event(EventType.MESSAGE_ARRIVED)
    
    def handle_event(self, event: Event) -> bool:
        """Handle incoming events"""
        if event.get_type() == EventType.MESSAGE_ARRIVED:
            data = event.get_data()
            topic: LaprasTopic = data[0]
            payload: bytes = data[1]
            
            # Check if this is a functionality message
            if topic.get_message_type() != MessageType.FUNCTIONALITY:
                return True
            
            # Parse the functionality invocation
            functionality_invocation = FunctionalityInvocation.from_payload(payload)
            if not functionality_invocation:
                return True
            
            # Invoke the functionality
            self.invoke_functionality(functionality_invocation)
            return True
        
        return False
    
    def get_functionality_signature(self, functionality_name: str) -> Optional[FunctionalitySignature]:
        """Get the signature for a functionality"""
        return self.functionality_signature_map.get(functionality_name)
    
    def list_functionality_signatures(self) -> List[FunctionalitySignature]:
        """List all registered functionality signatures"""
        return list(self.functionality_signature_map.values())
    
    def register_functionality(self, functionality_name: str, function: Callable, signature: Optional[FunctionalitySignature] = None):
        """Register a functionality with the system"""
        self.functionality_function_map[functionality_name] = function
        
        if not signature:
            # Create a minimal signature if none provided
            signature = FunctionalitySignature(functionality_name, [], "")
        
        self.functionality_signature_map[functionality_name] = signature
        
        self.logger.debug(f"Functionality {functionality_name} has been registered")
        
        # Subscribe to MQTT topic for this functionality
        topic = LaprasTopic(None, MessageType.FUNCTIONALITY, functionality_name)
        self.mqtt_communicator.subscribe_topic(topic)
    
    def register_method_functionality(self, functionality_name: str, method, instance):
        """Register a method as a functionality"""
        # Get signature from method
        signature = FunctionalitySignature.from_method_with_name(functionality_name, method)
        
        # Create function that invokes the method
        def invoke_method(arguments):
            try:
                return method(instance, *arguments)
            except Exception as e:
                self.logger.error(f"Error executing functionality {functionality_name}: {e}")
                return None
        
        # Register the functionality
        self.register_functionality(functionality_name, invoke_method, signature)
    
    def invoke_functionality(self, functionality_invocation: FunctionalityInvocation):
        """Invoke a functionality locally"""
        functionality_name = functionality_invocation.get_name()
        function = self.functionality_function_map.get(functionality_name)
        
        if not function:
            self.logger.warning(f"No functionality registered with name: {functionality_name}")
            return
        
        # Process arguments according to signature
        signature = self.functionality_signature_map.get(functionality_name)
        arguments = self._process_arguments(functionality_invocation.get_arguments(), signature)
        
        # Submit the functionality for execution
        self.functionality_executor.submit(self._execute_functionality, functionality_name, function, arguments)
    
    def invoke_functionality_by_name(self, functionality_name: str, arguments: List[Any]):
        """Invoke a functionality by name with the given arguments"""
        functionality_invocation = FunctionalityInvocation(
            functionality_name, 
            arguments, 
            int(time.time() * 1000), 
            self.agent_name
        )
        self.invoke_functionality(functionality_invocation)
    
    def invoke_remote_functionality(self, functionality_name: str, arguments: List[Any]):
        """Invoke a functionality on a remote agent via MQTT"""
        functionality_invocation = FunctionalityInvocation(
            functionality_name, 
            arguments, 
            int(time.time() * 1000), 
            self.agent_name
        )
        
        # Publish to MQTT
        topic = LaprasTopic(None, MessageType.FUNCTIONALITY, functionality_name)
        self.mqtt_communicator.publish(topic, functionality_invocation)
    
    def _execute_functionality(self, functionality_name: str, function: Callable, arguments: List[Any]):
        """Execute a functionality function with the given arguments"""
        self.logger.debug(f"Invoking functionality {functionality_name}")
        try:
            function(arguments)
        except Exception as e:
            self.logger.error(f"Error executing functionality {functionality_name}: {e}")
    
    def _process_arguments(self, arguments: List[Any], signature: FunctionalitySignature) -> List[Any]:
        """Process and validate arguments against a functionality signature"""
        if arguments is None:
            if signature and len(signature.get_parameters()) > 0:
                raise ValueError("Arguments expected but none provided")
            return []
        
        if signature and len(arguments) != len(signature.get_parameters()):
            raise ValueError(f"Expected {len(signature.get_parameters())} arguments, got {len(arguments)}")
        
        # If no signature or no parameters, return arguments as is
        if not signature or not signature.get_parameters():
            return arguments
        
        result = []
        for i, param_sig in enumerate(signature.get_parameters()):
            if param_sig.get_type() == DataType.FILE:
                # Handle file type: decode base64 and save to temp file
                try:
                    file_base64 = str(arguments[i]).replace('\n', '')
                    file_bytes = base64.b64decode(file_base64)
                    
                    # Create temp file
                    fd, path = tempfile.mkstemp(prefix="lapras_")
                    with os.fdopen(fd, 'wb') as tmp:
                        tmp.write(file_bytes)
                    
                    result.append(path)
                except Exception as e:
                    self.logger.error(f"Error processing file argument: {e}")
                    result.append(None)
            else:
                # Pass other arguments through directly
                result.append(arguments[i])
        
        return result

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/functionality/FunctionalityExecutor.java
package kr.ac.kaist.cdsn.lapras.functionality;

import kr.ac.kaist.cdsn.lapras.Agent;
import kr.ac.kaist.cdsn.lapras.Component;
import kr.ac.kaist.cdsn.lapras.communicator.LaprasTopic;
import kr.ac.kaist.cdsn.lapras.communicator.MessageType;
import kr.ac.kaist.cdsn.lapras.communicator.MqttCommunicator;
import kr.ac.kaist.cdsn.lapras.event.Event;
import kr.ac.kaist.cdsn.lapras.event.EventDispatcher;
import kr.ac.kaist.cdsn.lapras.event.EventType;
import kr.ac.kaist.cdsn.lapras.util.DataType;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * Created by Daekeun Lee on 2016-11-15.
 */
public class FunctionalityExecutor extends Component {
    private static Logger LOGGER = LoggerFactory.getLogger(FunctionalityExecutor.class);

    private final ExecutorService functionalityExecutor = Executors.newSingleThreadExecutor();
    private final ConcurrentMap<String, Consumer<Object[]>> functionalityFunctionMap = new ConcurrentHashMap<>();
    private final ConcurrentMap<String, FunctionalitySignature> functionalitySignatureMap = new ConcurrentHashMap<>();

    private final MqttCommunicator mqttCommunicator;
    private final String agentName;

    public FunctionalityExecutor(EventDispatcher eventDispatcher, Agent agent) {
        super(eventDispatcher, agent);

        mqttCommunicator = agent.getMqttCommunicator();
        agentName = agent.getAgentConfig().getAgentName();
    }

    @Override
    protected void subscribeEvents() {
        subscribeEvent(EventType.MESSAGE_ARRIVED);
    }

    @Override
    protected boolean handleEvent(Event event) {
        switch(event.getType()) {
            case MESSAGE_ARRIVED:
                LaprasTopic topic = (LaprasTopic) ((Object[]) event.getData())[0];
                byte[] payload = (byte[]) ((Object[]) event.getData())[1];
                if(!topic.getMessageType().equals(MessageType.FUNCTIONALITY)) return true;
                FunctionalityInvocation functionalityInvocation = FunctionalityInvocation.fromPayload(payload);
                if(functionalityInvocation == null) return true;
                invokeFunctionality(functionalityInvocation);
                return true;
        }
        return false;
    }

    public FunctionalitySignature getFunctionalitySignature(String functionalityName) {
        return functionalitySignatureMap.get(functionalityName);
    }

    public List<FunctionalitySignature> listFunctionalitySignatures() {
        return functionalitySignatureMap.entrySet().stream().map(entry->entry.getValue()).collect(Collectors.toList());
    }

    public void registerFunctionality(String functionalityName, Consumer<Object[]> function) {
        functionalityFunctionMap.put(functionalityName, function);

        FunctionalitySignature signature = new FunctionalitySignature(functionalityName, new ArrayList<>(), "");
        functionalitySignatureMap.put(functionalityName, signature);

        LOGGER.debug("Functionality {} has been registered", functionalityName);

        LaprasTopic topic = new LaprasTopic(null, MessageType.FUNCTIONALITY, functionalityName);
        mqttCommunicator.subscribeTopic(topic);
    }

    public void registerFunctionality(String functionalityName, Method method, Object instance) {
        FunctionalitySignature signature = FunctionalitySignature.fromMethodWithName(functionalityName, method);
        registerFunctionality(functionalityName, (arguments)->{
            try {
                method.invoke(instance, arguments);
            } catch (IllegalAccessException e) {
                LOGGER.error("Cannot invoke the method {} of {}", method.getName(), instance.toString(), e);
            } catch (InvocationTargetException e) {
                LOGGER.error("An error occurred while executing functionality {}", functionalityName, e);
            }
        });
        functionalitySignatureMap.put(functionalityName, signature);
    }

    public void invokeFunctionality(String functionalityName, Object[] arguments) {
        FunctionalityInvocation functionalityInvocation = new FunctionalityInvocation(functionalityName, arguments, System.currentTimeMillis(), agentName);
        invokeFunctionality(functionalityInvocation);
    }

    public void invokeRemoteFunctionality(String functionalityName, Object[] arguments) {
        FunctionalityInvocation functionalityInvocation = new FunctionalityInvocation(functionalityName, arguments, System.currentTimeMillis(), agentName);
        LaprasTopic topic = new LaprasTopic(null, MessageType.FUNCTIONALITY, functionalityName);
        mqttCommunicator.publish(topic, functionalityInvocation);
    }

    private void invokeFunctionality(FunctionalityInvocation functionalityInvocation) {
        final String functionalityName = functionalityInvocation.getName();
        final Consumer<Object[]> function = functionalityFunctionMap.get(functionalityName);
        final Object[] arguments = processArguments(functionalityInvocation.getArguments(), functionalitySignatureMap.get(functionalityName));
        functionalityExecutor.execute(() -> {
            LOGGER.debug("Invoking functionality {}", functionalityName);
            function.accept(arguments);
        });
    }

    private Object[] processArguments(Object[] arguments, FunctionalitySignature signature) {
        if(arguments == null) {
            if(signature.getParameters().size() != 0) {
                throw new IllegalArgumentException("Number of arguments don't agree");
            }
            return null;
        } else if(arguments.length != signature.getParameters().size()) {
            throw new IllegalArgumentException("Number of arguments don't agree");
        }

        Object[] result = new Object[arguments.length];
        for (int i = 0; i < signature.getParameters().size(); i++) {
            FunctionalitySignature.ParameterSignature parameterSignature = signature.getParameters().get(i);
            if(parameterSignature.getType() == DataType.FILE) {
                String fileBase64 = ((String) arguments[i]).replaceAll("\n", "");
                byte[] bytes = Base64.getDecoder().decode(fileBase64);
                try {
                    File tempfile = File.createTempFile("lapras", null);
                    FileUtils.writeByteArrayToFile(tempfile, bytes);
                    String path = tempfile.getAbsolutePath();
                    result[i] = new File(path);
                } catch (IOException e) {
                    LOGGER.error("Error while saving to temp file", e);
                    result[i] = null;
                    continue;
                }
            } else {
                result[i] = arguments[i];
            }
        }
        return result;
    }
}
